Sources/OpenAPIRuntime/EventStreams/JSONSequenceDecoding.swift

```diff
@@ -0,0 +1,236 @@
+    /// - Parameters:
+    ///   - eventType: The type to decode the JSON event into.
+    ///   - decoder: The JSON decoder to use.
+    /// - Returns: A sequence that provides the decoded JSON events.
+    public func asDecodedJSONSequence<Event: Decodable>(
+        of eventType: Event.Type = Event.self,
+        decoder: JSONDecoder = .init()
+    ) -> AsyncThrowingMapSequence<JSONSequenceDeserializationSequence<Self>, Event> {
+        JSONSequenceDeserializationSequence(upstream: self)
+            .map { line in try decoder.decode(Event.self, from: Data(line)) }
+    }
+}
+
+extension JSONSequenceDeserializationSequence.Iterator {
+
+    /// A state machine representing the JSON Lines deserializer.
+    struct StateMachine {
+
+        /// The possible states of the state machine.
+        enum State: Hashable {
+
+            /// Has not yet fully parsed the initial boundary.
+            case initial(buffer: [UInt8])
+
+            /// Is parsing a line, waiting for the end newline.
+            case parsingLine(buffer: [UInt8])
+
+            /// Finished, the terminal state.
+            case finished
+
+            /// Helper state to avoid copy-on-write copies.
+            case mutating
+        }
+
+        /// The current state of the state machine.
+        private(set) var state: State
+
+        /// Creates a new state machine.
+        init() { self.state = .initial(buffer: []) }
+
+        /// An error returned by the state machine.
+        enum ActionError {
+
+            /// The initial boundary `<RS>` was not found.
+            case missingInitialRS
+        }
+
+        /// An action returned by the `next` method.
+        enum NextAction {
+
+            /// Return nil to the caller, no more bytes.
+            case returnNil
+
+            /// Emit a full line.
+            case emitLine(ArraySlice<UInt8>)
+
+            /// Emit an error.
+            case emitError(ActionError)
+
+            /// The line is not complete yet, needs more bytes.
+            case needsMore
+
+            /// Rerun the parsing loop.
+            case noop
+        }
+
+        /// Read the next line parsed from upstream bytes.
+        /// - Returns: An action to perform.
+        mutating func next() -> NextAction {
+            switch state {
+            case .initial(var buffer):
+                guard !buffer.isEmpty else { return .needsMore }
+                guard buffer.first! == ASCII.rs else { return .emitError(.missingInitialRS) }
+                state = .mutating
+                buffer.removeFirst()
+                state = .parsingLine(buffer: buffer)
+                return .noop
+            case .parsingLine(var buffer):
+                state = .mutating
+                guard let indexOfRecordSeparator = buffer.firstIndex(of: ASCII.rs) else {
+                    state = .parsingLine(buffer: buffer)
+                    return .needsMore
+                }
```

Comment:
Not a blocker, but I'm wondering if we can do better here with respect to performance. Specifically, this will call `firstIndex(of:)` every time we receive new bytes to find the record separator, which will be O(n) in the current size of the buffer and will search for the record separator through portions of the buffer that may have been searched last time.

A pathological example would be a sequence of one-byte chunks, in which case it will be quadratic in the size of the line, but it needn't be.

It can be linear if we:

1. Augment the `.parsingLine(buffer:)` state to store the current `endIndex` of the buffer before returning `.needsMore` or something similar.
2. Use this with a new utility function `firstIndex(of:from:)`, which would look for the first index from a starting index.

This comment might apply to the other sequences in this PR too.

Reply:
It'd apply to other sequences _outside_ this PR as well, the multipart ones. Could you file that as an issue? I'm not closed to it and I even tried it briefly, but it exploded the complexity a bit too much for my liking, so I backed it out. But it was also during the prototyping phase, where everything was in flux, so I'd be happy to investigate doing this in all the parsers in the runtime library as a standalone piece of work.

Reply:
Filed an issue to track this separately: https://github.com/apple/swift-openapi-runtime/issues/93#issue-2070219477